llpc/util/llpcElfWriter.cpp

```diff
@@ -1118,6 +1205,152 @@ void ElfWriter<Elf>::Reinitialize()
+        }
+        else
+        {
+            return (resource->offsetInDwords) * sizeof(uint32_t);
+        }
+    }
+    return InvalidValue;
+}
+
+// =====================================================================================================================
+// Fixup a descriptor offset relocation (doff_x_y_t symbol) in the ELF with an actual offset value.
+bool GetDescriptorOffsetRelocationValue(
+    Context* pContext,               // [in] Pipeline compilation context
+    RelocationEntry relocEntry,      // The relocation entry to fixup
+    bool isGraphicsPipeline,         // Whether we are processing a compute or graphics pipeline
+    uint32_t* pValue)                 // [out] The value of the relocation
+{
+    size_t idx = 0;
+    const char* pRelocName = relocEntry.name + 5;
+    uint32_t descSet = std::stoi(pRelocName, &idx);
+    pRelocName += idx + 1;
+    idx = 0;
+    uint32_t binding = std::stoi(pRelocName, &idx);
+    pRelocName += idx + 1;
+    ResourceMappingNodeType type = ResourceMappingNodeType::Unknown;
+    switch (*pRelocName)
+    {
+    case 's':
+        type = ResourceMappingNodeType::DescriptorSampler;
+        break;
+    case 'r':
+        type = ResourceMappingNodeType::DescriptorResource;
+        break;
+    case 'b':
+        type = ResourceMappingNodeType::DescriptorBuffer;
+        break;
+    }
+
+    if (isGraphicsPipeline)
+    {
+        auto pPipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->vs.pUserDataNodes,
+                                              pPipelineInfo->vs.userDataNodeCount);
+        if (*pValue == InvalidValue)
+        {
+            *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->fs.pUserDataNodes,
+                                                  pPipelineInfo->fs.userDataNodeCount);
+        }
+    }
+    else
+    {
+        auto pPipelineInfo = reinterpret_cast<const ComputePipelineBuildInfo*>(pContext->GetPipelineBuildInfo());
+        *pValue = GetDescriptorResourceOffset(descSet, binding, type, pPipelineInfo->cs.pUserDataNodes,
+                                              pPipelineInfo->cs.userDataNodeCount);
+    }
+
+    return *pValue != InvalidValue;
+}
+
+// =====================================================================================================================
+// Get the value of a relocation symbol. Returns true if success, and false if the symbol is unknown.
+bool GetRelocationSymbolValue(
+    Context* pContext,       // [in] Pipeline compilation context
+    RelocationEntry reloc,   // The relocation entry
+    bool isGraphicsPipeline, // Whether we are processing a graphics pipeline
+    uint32_t* pValue)         // [out] the value of the relocation symbol
+{
+    assert(reloc.type == R_AMDGPU_ABS32 && "can only handle R_AMDGPU_ABS32 typed relocations.");
```

Comment:
Please move this assertion into FixUpRelocations instead: GetRelocationSymbolValue is the place that knows how to get a symbol value, FixUpRelocations is the place that knows what to do with it.